pkg/transport/generic_transport_controller.go

```diff
@@ -0,0 +1,512 @@
+	})
+	// Set up event handlers for when WrappedObject resources change. The handlers will lookup the origin PlacementDecision
+	// of the given WrappedObject and enqueue that PlacementDecision object for processing.
+	// This way, we don't need to implement custom logic for handling WrappedObject resources. More info on this pattern:
+	// https://github.com/kubernetes/community/blob/8cafef897a22026d42f5e5bb3f104febe7e29830/contributors/devel/controllers.md
+	wrappedObjectGenericInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    transportController.handleWrappedObject,
+		UpdateFunc: func(_, new interface{}) { transportController.handleWrappedObject(new) },
+		DeleteFunc: transportController.handleWrappedObject,
+	})
+	dynamicInformer.Start(ctx.Done())
+
+	return transportController, nil
+}
+
+func convertObjectToUnstructured(object runtime.Object) (*unstructured.Unstructured, error) {
+	unstructuredObject, err := runtime.DefaultUnstructuredConverter.ToUnstructured(object)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert given object to unstructured - %w", err)
+	}
+	return &unstructured.Unstructured{Object: unstructuredObject}, nil
+}
+
+func getGvrFromWrappedObject(restConfig *rest.Config, wrappedObject runtime.Object) (*schema.GroupVersionResource, error) {
+	unstructuredWrappedObject, err := convertObjectToUnstructured(wrappedObject)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert wrapped object to unstructured - %w", err)
+	}
+
+	gvk := unstructuredWrappedObject.GroupVersionKind()
+	clientset, err := kubernetes.NewForConfig(restConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create k8s clientset for given config - %w", err)
+	}
+	mapper := restmapper.NewDeferredDiscoveryRESTMapper(cacheddiscovery.NewMemCacheClient(clientset.Discovery()))
+
+	restMapping, err := mapper.RESTMapping(gvk.GroupKind(), gvk.Version)
+	if err != nil {
+		return nil, fmt.Errorf("failed to discover GroupVersionResource from given GroupVersionKind - %w", err)
+	}
+
+	return &(restMapping.Resource), nil
+}
+
+type genericTransportController struct {
+	logger logr.Logger
+
+	placementDecisionLister         edgev1alpha1listers.PlacementDecisionLister
+	placementDecisionInformerSynced cache.InformerSynced
+	wrappedObjectInformerSynced     cache.InformerSynced
+
+	// workqueue is a rate limited work queue.
+	// This is used to queue work to be processed instead of performing it as soon as a change happens.
+	// This means we can ensure we only process a fixed amount of resources at a time, and makes it
+	// easy to ensure we are never processing the same item simultaneously in two different workers.
+	workqueue workqueue.RateLimitingInterface
+
+	//transport is a specific implementation for the transport interface.
+	transport        Transport
+	transportClient  dynamic.Interface // dynamic client to transport wrapped object. since object kind is unknown during complilation, we use dynamic
```

Comment:
Have you considered using golang generics?

Reply:
no. using dynamic client here for a given GVR is very straight forward and simple.
don't think we need generics here.